market/contracts.py

```diff
@@ -420,6 +430,112 @@ def accept_order_confirmation(self, ws, confirmation_json=None):
         except Exception:
             return False
 
+    def add_receipt(self,
+                    received,
+                    feedback=None,
+                    quality=None,
+                    description=None,
+                    delivery_time=None,
+                    customer_service=None,
+                    review="",
+                    dispute=False,
+                    claim=None,
+                    payout=True):
+
+        """
+        Add the final piece of the contract that appends the review and payout transaction.
+        """
+        receipt_json = {
+            "buyer_receipt": {
+                "receipt": {
+                    "ref_hash": digest(json.dumps(self.contract, indent=4)).encode("hex"),
+                    "listing": {
+                        "received": received,
+                        "listing_hash": self.contract["buyer_order"]["order"]["ref_hash"]
+                    },
+                    "dispute": {
+                        "dispute": dispute
+                    }
+                }
+            }
+        }
+        if None in (feedback, quality, description, delivery_time, customer_service):
+            rating = False
+        else:
+            rating = True
+            receipt_json["buyer_receipt"]["receipt"]["rating"] = {}
+            receipt_json["buyer_receipt"]["receipt"]["rating"]["feedback"] = feedback
+            receipt_json["buyer_receipt"]["receipt"]["rating"]["quality"] = quality
+            receipt_json["buyer_receipt"]["receipt"]["rating"]["description"] = description
+            receipt_json["buyer_receipt"]["receipt"]["rating"]["delivery_time"] = delivery_time
+            receipt_json["buyer_receipt"]["receipt"]["rating"]["customer_service"] = customer_service
+            receipt_json["buyer_receipt"]["receipt"]["rating"]["review"] = review
+        if payout:
+            vendor_guid = self.contract["vendor_offer"]["listing"]["id"]["guid"]
+            contract_id = self.contract["buyer_order"]["order"]["ref_hash"]
+            order_id = self.contract["vendor_order_confirmation"]["invoice"]["ref_hash"]
+            outpoints = pickle.loads(self.db.Purchases().get_outpoint(order_id))
```

Comment:
also a little out of order as I see the use of JSON and protobuf everywhere else.